File: src/lib/rubeno/lib/base_suite.rb
module Rubeno
  class BaseSuite
    attr_accessor :name, :givens, :store, :test_resource_configuration, :index, :failed, :fails, :artifacts, :parent
    
    def initialize(name, givens = {})
      @name = name
      @givens = givens
      @artifacts = []
      @failed = false
      @fails = 0
    end
    
    def add_artifact(path)
      normalized_path = path.gsub('\\', '/')
      @artifacts << normalized_path
    end
    
    def features
      features = []
      seen = {}
      @givens.each_value do |given|
        given.features.each do |feature|
          unless seen[feature]
            features << feature
            seen[feature] = true
          end
        end
      end
      features
    end
    
    def to_obj
      {
        name: @name,
        givens: @givens.values.map(&:to_obj),
        fails: @fails,
        failed: @failed,
        features: features,
        artifacts: @artifacts
      }
    end
    
    def setup(s, tr, artifactory = nil)
      s
    end
    
    def assert_that(t)
      !!t
    end
    
    def after_all(store, artifactory = nil)
      store
    end
    
    def run(input_val, test_resource_configuration)
      @test_resource_configuration = test_resource_configuration
      
      # Create artifactory for suite setup
      suite_artifactory = nil
      if @parent && @parent.respond_to?(:create_artifactory)
        suite_artifactory = @parent.create_artifactory(suite_index: @index)
      end
      
      subject = setup(input_val, test_resource_configuration, suite_artifactory)
      
      @fails = 0
      @failed = false
      
      @givens.each do |g_key, g|
        begin
          # Create artifactory for the given
          given_artifactory = nil
          if @parent && @parent.respond_to?(:create_artifactory)
            given_artifactory = @parent.create_artifactory(given_key: g_key, suite_index: @index)
          end
          
          # Set parent reference if not already set
          g._parent = @parent if g.respond_to?(:_parent=)
          
          @store = g.give(
            subject,
            g_key,
            test_resource_configuration,
            method(:assert_that),
            given_artifactory,
            @index
          )
          @fails += g.fails if g.fails && g.fails > 0
        rescue => e
          @failed = true
          @fails += 1
          puts "Error in given #{g_key}: #{e.message}"
        end
      end
      
      @failed = true if @fails > 0
      
      begin
        after_all(@store, suite_artifactory)
      rescue => e
        puts "Error in after_all: #{e.message}"
      end
      
      self
    end
  end
end

File: src/lib/kafe/src/main/java/kafe/BaseSuite.java
package kafe;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

public abstract class BaseSuite {
    public String name;
    public Map<String, BaseGiven> givens;
    public Object store;
    public ITTestResourceConfiguration testResourceConfiguration;
    public int index;
    public boolean failed;
    public int fails;
    public List<String> artifacts;
    public Object parent; // Reference to parent Kafe instance
    
    public BaseSuite(String name, Map<String, BaseGiven> givens) {
        this.name = name;
        this.givens = givens != null ? givens : new HashMap<>();
        this.artifacts = new ArrayList<>();
        this.failed = false;
        this.fails = 0;
        this.index = 0;
    }
    
    public void addArtifact(String path) {
        String normalizedPath = path.replace('\\', '/');
        this.artifacts.add(normalizedPath);
    }
    
    public List<String> features() {
        List<String> features = new ArrayList<>();
        Map<String, Boolean> seen = new HashMap<>();
        
        for (BaseGiven given : givens.values()) {
            if (given.features != null) {
                for (Object featureObj : given.features) {
                    String feature = featureObj.toString();
                    if (!seen.containsKey(feature)) {
                        features.add(feature);
                        seen.put(feature, true);
                    }
                }
            }
        }
        return features;
    }
    
    public Map<String, Object> toObj() {
        List<Map<String, Object>> givenObjs = new ArrayList<>();
        for (BaseGiven given : givens.values()) {
            givenObjs.add(given.toObj());
        }
        
        Map<String, Object> obj = new HashMap<>();
        obj.put("name", name);
        obj.put("givens", givenObjs);
        obj.put("fails", fails);
        obj.put("failed", failed);
        obj.put("features", features());
        obj.put("artifacts", artifacts);
        return obj;
    }
    
    public abstract Object setup(Object s, ITTestResourceConfiguration tr);
    
    public abstract boolean assertThat(Object t);
    
    public abstract Object afterAll(Object store);
    
    public Object run(Object input, ITTestResourceConfiguration testResourceConfiguration) {
        this.testResourceConfiguration = testResourceConfiguration;
        
        // Create artifactory for suite
        Object suiteArtifactory = createSuiteArtifactory();
        
        // Call setup with artifactory
        Object subject;
        try {
            subject = setup(input, testResourceConfiguration);
        } catch (Exception e) {
            System.err.println("Error in suite setup: " + e.getMessage());
            this.failed = true;
            this.fails++;
            return this;
        }
        
        fails = 0;
        failed = false;
        
        for (Map.Entry<String, BaseGiven> entry : givens.entrySet()) {
            String gKey = entry.getKey();
            BaseGiven g = entry.getValue();
            try {
                // Create artifactory for given
                Object givenArtifactory = createGivenArtifactory(gKey);
                
                store = g.give(subject, gKey, testResourceConfiguration, this::assertThat,
                              givenArtifactory, // Pass artifactory as Object
                              this.index);
                if (g.failed) {
                    fails += g.fails;
                }
            } catch (Exception e) {
                failed = true;
                fails++;
                System.err.println("Error in given " + gKey + ": " + e.getMessage());
                e.printStackTrace();
            }
        }
        
        if (fails > 0) {
            failed = true;
        }
        
        try {
            // Create artifactory for afterAll
            Object afterAllArtifactory = createSuiteArtifactory();
            afterAll(store);
        } catch (Exception e) {
            System.err.println("Error in after_all: " + e.getMessage());
        }
        
        return this;
    }
    
    private Object createSuiteArtifactory() {
        if (parent != null) {
            try {
                java.lang.reflect.Method createArtifactory = parent.getClass().getMethod("createArtifactory", Map.class);
                Map<String, Object> context = new HashMap<>();
                context.put("suiteIndex", this.index);
                return createArtifactory.invoke(parent, context);
            } catch (Exception e) {
                System.err.println("Could not create suite artifactory: " + e.getMessage());
            }
        }
        return null;
    }
    
    private Object createGivenArtifactory(String givenKey) {
        if (parent != null) {
            try {
                java.lang.reflect.Method createArtifactory = parent.getClass().getMethod("createArtifactory", Map.class);
                Map<String, Object> context = new HashMap<>();
                context.put("givenKey", givenKey);
                context.put("suiteIndex", this.index);
                return createArtifactory.invoke(parent, context);
            } catch (Exception e) {
                System.err.println("Could not create given artifactory: " + e.getMessage());
            }
        }
        return null;
    }
    
    // Set parent reference
    public void setParent(Object parent) {
        this.parent = parent;
    }
}
